openlcsd/flow/tasks.py

```diff
@@ -264,24 +284,39 @@ def download_package_archive(context, engine):
     tmp_dir = tempfile.mkdtemp(
         prefix='download_', dir=context.get('tmp_root_dir'))
     config = context.get('config')
-    build = context.get('build')
     koji_connector = KojiConnector(config)
-    build_id = build.get('id')
-    engine.logger.info('Start to download package source from Brew/Koji...')
+    component = context.get('component')
+    if build := context.get('build'):
+        build_id = build.get('id')
+    else:
+        software_build = component.get('software_build') if component else None
+        build_id = software_build.get('build_id')
+    engine.logger.info('Start to download package source...')
     try:
-        koji_connector.download_build_source(build_id, dest_dir=tmp_dir)
+        if build_id:
+            koji_connector.download_build_source(build_id, dest_dir=tmp_dir)
+        else:
+            # For remote source, we should use 'download_url' to
+            # download pacakge archive
+            if download_url := component.get('download_url'):
+                cmd = f'wget -q --show-progress {download_url}'
```

Comment:
`wget` can not work for all remote sources, the `download_url` is not always a source tarball.
It should work for download_url for rubygems, pypi or npm remote sources:
```
https://rubygems.org/downloads/3scale-api-1.4.0.gem
https://pypi.io/packages/source/a/aiodns/aiodns-3.0.0.tar.gz
http://registry.npmjs.org/0-dns/-/0-dns-1.0.0.tgz  
```
But can not work for that for maven, cargo or golang(maybe some others), their `download_url` examples in Corgi:
```
https://repo1.maven.org/maven2/a/a/0.1-20130115.024354-82/
https://crates.io/api/v1/crates/aardvark-dns/1.4.0/download
https://pkg.go.dev/4d63.com/gochecknoglobals/checknoglobals@v0.1.0
```

Reply:
Yes, I have asked Corgi, waiting for the response. And talked with @rhjzhao, if we cannot get the source, we will raise an exception, so that it will not block our progress.

Reply:
I don't think so, this is not blocking the progress but part of the process, raising exceptions is unreasonable. If OLCS has to get the source from the URL and the URL can only be provided this way, OLCS(/Corgi) has to deal with it, otherwise, the exception would be everywhere for containers with such remote sources, thus we can not say OLCS could support these types of remote sources, because it can not even get its source. Anyway, I'd like to see if any feedback from Corgi team about this. Or you could split the task.

Reply:
Here is something needs to be clarified, see feedback from Jason about `download_url`:

> I think you should limit the types you try to use download_url to find source to remote_source types (GOLANG, NPM, PYPI, and GEM). For components of type RPM (arch=src) and OCI (arch=noarch) you should not try to use download_url as the source for those will be better obtained from SoftwareBuild.source.

Reply:
Question: For a source container, OLCS is getting the rpm sources from Brew and remote sources(only NPM, PYPI and GEM are supported in this PR, GOLANG/CARGO/MAVEN will be tracked elsewhere) from the `download_url`, how about the top level source files in the container source image that doesn't belong to any of its components?

Reply:
What are `top-level source files in the container source image`? could you give an example?

Reply:
Besides the collection of component source artifacts, there are some other files for manifest / metadata in the container image. From Corgi:

> For the root Component objects OpenLCS is scanning (source RPMs and source container images), Corgi should always link to a "SoftwareBuild" object (which is the build data from Brew). That build data should have a "source" field with a git:// protocol link. You can download the source code using "git clone" with that link to do your scanning. 

and 

> For components of type RPM (arch=src) and OCI (arch=noarch) you should not try to use download_url as the source for those will be better obtained from SoftwareBuild.source.

Now OLCS is trying to get the container component source in a different way, I'm wondering where these top files are captured. Is this covered in this PR or has this been done somewhere else?

Reply:
This PR only covers the collection of component source artifacts, but I don't find any story to cover some other files for manifest/metadata in the container image.  so if it is necessary, we can create a new story at OLCS-354

Reply:
Good point. I think we should create a new issue to track the discussion with OSC and Corgi. For now, we will not consider this situation. That means we totally trust Corgi data during the integration.

> Besides the collection of component source artifacts, there are some other files for manifest / metadata in the container image.

Reply:
OK, I will file a new task to track it

Reply:
New task https://issues.redhat.com/browse/OLCS-452 to track it @qduanmu @rhjzhao

Reply:
This has nothing to do with trusting Corgi or not. There should be no missing source files for a container, no matter where you get its source.